Add --produce-test flag to generate test case from minifier

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101459

Given a file:

```
from math import inf
import torch
from torch import tensor, device
import torch.fx as fx
import torch._dynamo
from torch._dynamo.testing import rand_strided
from torch._dynamo.debug_utils import run_fwd_maybe_bwd

import torch._dynamo.config
import torch._inductor.config
import torch._functorch.config
torch._dynamo.config.debug_dir_root = '.'
torch._inductor.config.cpp.inject_relu_bug_TESTING_ONLY = 'compile_error'






from torch.nn import *
class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()



    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        sigmoid = torch.sigmoid(l_x_);  l_x_ = None
        ones = torch.ones(2)
        mul = torch.mul(sigmoid, ones);  sigmoid = ones = None
        relu = torch.relu(mul);  mul = None
        zeros = torch.zeros(2)
        add = torch.add(relu, zeros);  relu = zeros = None
        round_1 = torch.ops.aten.round(add);  add = None
        return (round_1,)


mod = Repro()

def load_args(reader):
    buf0 = reader.storage('b01abf7d0dfaf0ca12f8d2260c98f8c9b9d7bff9', 8)
    reader.tensor(buf0, (2,))  # L_x_
load_args._version = 0

if __name__ == '__main__':
    from torch._dynamo.repro.after_dynamo import run_repro
    run_repro(mod, load_args, accuracy=False, command='minify',
        save_dir='./run_2023_05_15_20_08_59_414896-pid_728954/minifier/checkpoints', autocast=False, backend='inductor')
```

If we run it with `python run_2023_05_15_20_08_59_414896-pid_728954/minifier/minifier_launcher.py minify --produce_test`

we will minify it down, and produce a test like:

```

from math import inf
import torch
from torch import tensor, device
import torch.fx as fxindent
import torch._dynamo
from torch._dynamo.testing import rand_strided
from torch._dynamo.debug_utils import run_fwd_maybe_bwd
from torch.nn import *
import torch._dynamo.test_case
import torch._dynamo.testing
import torch._dynamo.config
import torch._inductor.config
import torch._functorch.config



class GeneratedReproTestCase(torch._dynamo.test_case.TestCase):
    @torch._dynamo.config.patch('debug_dir_root', '.')
    @torch._inductor.config.patch('cpp.inject_relu_bug_TESTING_ONLY', 'compile_error')
    def test_generated_repro_case(self):


        class Repro(torch.nn.Module):
            def __init__(self):
                super().__init__()



            def forward(self, L_L_x_ : torch.Tensor):
                l_l_x_ = L_L_x_
                sigmoid = torch.sigmoid(l_l_x_);  l_l_x_ = None
                ones = torch.ones(2)
                mul = torch.mul(sigmoid, ones);  sigmoid = ones = None
                relu = torch.relu(mul);  mul = None
                zeros = torch.zeros(2)
                add = torch.add(relu, zeros);  relu = zeros = None
                round_1 = torch.ops.aten.round(add);  add = None
                return (round_1,)



        buf0 = torch.randn([2], dtype=torch.float32, device='cpu')  # L_L_x_ 

        # TODO(voz): Add dynamic, nopython, and all sorts of other configs
        optimized_result = torch._dynamo.optimize('inductor')(Repro())(buf0)
        eager_result = Repro()(buf0)
        self.assertEqual(optimized_result, eager_result)


if __name__ == '__main__':
    from torch._dynamo.test_case import run_tests

    run_tests()
```


cc @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov @desertfire

voznesenskym committed:
[WIP] Add produce_test flag to generate test case from minifier

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] Add --produce-test flag to generate test case from minifier"


Given a file:

```
from math import inf
import torch
from torch import tensor, device
import torch.fx as fx
import torch._dynamo
from torch._dynamo.testing import rand_strided
from torch._dynamo.debug_utils import run_fwd_maybe_bwd

import torch._dynamo.config
import torch._inductor.config
import torch._functorch.config
torch._dynamo.config.debug_dir_root = '.'
torch._inductor.config.cpp.inject_relu_bug_TESTING_ONLY = 'compile_error'






from torch.nn import *
class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()



    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        sigmoid = torch.sigmoid(l_x_);  l_x_ = None
        ones = torch.ones(2)
        mul = torch.mul(sigmoid, ones);  sigmoid = ones = None
        relu = torch.relu(mul);  mul = None
        zeros = torch.zeros(2)
        add = torch.add(relu, zeros);  relu = zeros = None
        round_1 = torch.ops.aten.round(add);  add = None
        return (round_1,)


mod = Repro()

def load_args(reader):
    buf0 = reader.storage('b01abf7d0dfaf0ca12f8d2260c98f8c9b9d7bff9', 8)
    reader.tensor(buf0, (2,))  # L_x_
load_args._version = 0

if __name__ == '__main__':
    from torch._dynamo.repro.after_dynamo import run_repro
    run_repro(mod, load_args, accuracy=False, command='minify',
        save_dir='./run_2023_05_15_20_08_59_414896-pid_728954/minifier/checkpoints', autocast=False, backend='inductor')
```

If we run it with `python run_2023_05_15_20_08_59_414896-pid_728954/minifier/minifier_launcher.py minify --produce_test`

we will minify it down, and produce a test like:

```

from math import inf
import torch
from torch import tensor, device
import torch.fx as fxindent
import torch._dynamo
from torch._dynamo.testing import rand_strided
from torch._dynamo.debug_utils import run_fwd_maybe_bwd
from torch.nn import *
import torch._dynamo.test_case
import torch._dynamo.testing

import torch._dynamo.config
import torch._inductor.config
import torch._functorch.config
torch._dynamo.config.debug_dir_root = '.'
torch._inductor.config.cpp.inject_relu_bug_TESTING_ONLY = 'compile_error'





class GeneratedReproTestCase(torch._dynamo.test_case.TestCase):
    def test_generated_repro_case(self):


        class Repro(torch.nn.Module):
            def __init__(self):
                super().__init__()



            def forward(self, mul):
                relu = torch.relu(mul);  mul = None
                return (relu,)


        mod = Repro()

        buf0 = torch.tensor([0.7310585975646973, 0.7310585975646973], dtype=torch.float32, device='cpu')  # mul 

        # TODO(voz): Add dynamic, nopython, and all sorts of other configs
        torch._dynamo.optimize('inductor')(mod)(buf0)

if __name__ == '__main__':
    from torch._dynamo.test_case import run_tests

    run_tests()

```


cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] Add --produce-test flag to generate test case from minifier"


Given a file:

```
from math import inf
import torch
from torch import tensor, device
import torch.fx as fx
import torch._dynamo
from torch._dynamo.testing import rand_strided
from torch._dynamo.debug_utils import run_fwd_maybe_bwd

import torch._dynamo.config
import torch._inductor.config
import torch._functorch.config
torch._dynamo.config.debug_dir_root = '.'
torch._inductor.config.cpp.inject_relu_bug_TESTING_ONLY = 'compile_error'






from torch.nn import *
class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()



    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        sigmoid = torch.sigmoid(l_x_);  l_x_ = None
        ones = torch.ones(2)
        mul = torch.mul(sigmoid, ones);  sigmoid = ones = None
        relu = torch.relu(mul);  mul = None
        zeros = torch.zeros(2)
        add = torch.add(relu, zeros);  relu = zeros = None
        round_1 = torch.ops.aten.round(add);  add = None
        return (round_1,)


mod = Repro()

def load_args(reader):
    buf0 = reader.storage('b01abf7d0dfaf0ca12f8d2260c98f8c9b9d7bff9', 8)
    reader.tensor(buf0, (2,))  # L_x_
load_args._version = 0

if __name__ == '__main__':
    from torch._dynamo.repro.after_dynamo import run_repro
    run_repro(mod, load_args, accuracy=False, command='minify',
        save_dir='./run_2023_05_15_20_08_59_414896-pid_728954/minifier/checkpoints', autocast=False, backend='inductor')
```

If we run it with `python run_2023_05_15_20_08_59_414896-pid_728954/minifier/minifier_launcher.py minify --produce_test`

we will minify it down, and produce a test like:

```

from math import inf
import torch
from torch import tensor, device
import torch.fx as fxindent
import torch._dynamo
from torch._dynamo.testing import rand_strided
from torch._dynamo.debug_utils import run_fwd_maybe_bwd
from torch.nn import *
import torch._dynamo.test_case
import torch._dynamo.testing

import torch._dynamo.config
import torch._inductor.config
import torch._functorch.config
torch._dynamo.config.debug_dir_root = '.'
torch._inductor.config.cpp.inject_relu_bug_TESTING_ONLY = 'compile_error'





class GeneratedReproTestCase(torch._dynamo.test_case.TestCase):
    def test_generated_repro_case(self):


        class Repro(torch.nn.Module):
            def __init__(self):
                super().__init__()



            def forward(self, mul):
                relu = torch.relu(mul);  mul = None
                return (relu,)


        mod = Repro()

        buf0 = torch.tensor([0.7310585975646973, 0.7310585975646973], dtype=torch.float32, device='cpu')  # mul 

        # TODO(voz): Add dynamic, nopython, and all sorts of other configs
        torch._dynamo.optimize('inductor')(mod)(buf0)

if __name__ == '__main__':
    from torch._dynamo.test_case import run_tests

    run_tests()

```


cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on torch/_dynamo/repro/after_dynamo.py:
```diff
@@ -497,6 +583,12 @@ def common_flags(parser):
             help="don't use any directory for saved inputs",
         )
+        parser.add_argument(
+            "--produce_test",
```
This is better as a mode, I think. Also, use of underscored flagname is inconsistent with the rest of the code

ezyang replied:
Because it is mutually exclusive with the other options (running, minifying)

voznesenskym replied:
Oh! I intended it to be additive. The test is a side effect, not an alternate mode. I think this offers higher value that way, and is easier to write because the minifier produces a runnable minifier which then repeats over and over as it... minifies.

ezyang commented on torch/_dynamo/debug_utils.py:
```diff
@@ -151,14 +151,14 @@ def can_convert_to_string(gm):
-            """
-            from torch.nn import *
+            f"""
+            {'from torch.nn import *' if imports else ''}
```
This seems a bit questionable

voznesenskym replied:
If you keep this here, you cannot nest this class. You can only have * imports top level.

ezyang commented on torch/_dynamo/debug_utils.py:
```diff
@@ -583,6 +583,7 @@ def symint(self, val):
 class InputWriter:
     def __init__(self, save_dir, *, stable_hash=False):
         self._lines = []
+        self._storage_names = []
```
unused?

voznesenskym replied:
Used, outside the class. I'll change it to be accessed by func or at least not private.

ezyang commented on torch/_dynamo/repro/after_dynamo.py:
```diff
@@ -184,7 +185,72 @@ def generate_dynamo_fx_repro_string(
+{extra_imports}
+
+class GeneratedReproTestCase(torch._dynamo.test_case.TestCase):
+    def test_generated_repro_case(self):
```
Don't you want the configs patched at the test case, rather than at the top of the file. Since ostensibly you're going to put the test case into one of the preexisting test case files?

voznesenskym replied:
>  Since ostensibly you're going to put the test case into one of the preexisting test case files?

No, I assumed you would just generate a test here for the person to copy/paste. I do not want to pollute or change existing code as part of this, that could make for ugly commits or unintended writes.

voznesenskym replied:
Oh you mean like, generate @patch config statements? Yeah, we can do that.

ezyang commented on torch/_dynamo/debug_utils.py:
```diff
@@ -665,3 +673,30 @@ def symint(self, name, val) -> None:
+        self._storage_names.append(f"{storage}")
+        self._lines.append(
+            f"{storage} = torch.randn({list(t.size())}, dtype={t.dtype}, device='{t.device.type}')  # {name} \n"
+        )
```
This isn't a faithful representation of the original generated test, because you are dropping the aliasing relationships entirely this way.

voznesenskym replied:
What should I be doing here?

ezyang replied:
That is a bit of a complicated question.

If the aliasing doesn't matter, it is simpler to write out a test case that isn't doing anything fancy, as you are doing here. Similarly, if there is no aliasing, you can't do much better than the logic here.

However, if the aliasing does matter, you now have a test case that doesn't repro the original problem. Instead, you need to output a sequence of operations that will produce the two tensors with the same aliasing as before. One simple to generate strategy is to use as_strided or set_ to construct the views off of a storage buffer. However, that too will often not result in idiomatic test code; if you could get the alias just by view(...)ing a tensor, that would be better!

The version of this code with *craft* has a number of candidate ways to reconstruct tensors, and applies the most idiomatic one that applies. It also teaches the minifier to attempt shrinking the test tensors, in case smaller size inputs are also able to trigger the test.

The easy to land version of this code probably intentionally fails to alias and writes a comment to the file saying that this is what happened, and has an option to also do things faithfully (but ugly).

voznesenskym committed:
Update on "[WIP] Add --produce-test flag to generate test case from minifier"


Given a file:

```
from math import inf
import torch
from torch import tensor, device
import torch.fx as fx
import torch._dynamo
from torch._dynamo.testing import rand_strided
from torch._dynamo.debug_utils import run_fwd_maybe_bwd

import torch._dynamo.config
import torch._inductor.config
import torch._functorch.config
torch._dynamo.config.debug_dir_root = '.'
torch._inductor.config.cpp.inject_relu_bug_TESTING_ONLY = 'compile_error'






from torch.nn import *
class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()



    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        sigmoid = torch.sigmoid(l_x_);  l_x_ = None
        ones = torch.ones(2)
        mul = torch.mul(sigmoid, ones);  sigmoid = ones = None
        relu = torch.relu(mul);  mul = None
        zeros = torch.zeros(2)
        add = torch.add(relu, zeros);  relu = zeros = None
        round_1 = torch.ops.aten.round(add);  add = None
        return (round_1,)


mod = Repro()

def load_args(reader):
    buf0 = reader.storage('b01abf7d0dfaf0ca12f8d2260c98f8c9b9d7bff9', 8)
    reader.tensor(buf0, (2,))  # L_x_
load_args._version = 0

if __name__ == '__main__':
    from torch._dynamo.repro.after_dynamo import run_repro
    run_repro(mod, load_args, accuracy=False, command='minify',
        save_dir='./run_2023_05_15_20_08_59_414896-pid_728954/minifier/checkpoints', autocast=False, backend='inductor')
```

If we run it with `python run_2023_05_15_20_08_59_414896-pid_728954/minifier/minifier_launcher.py minify --produce_test`

we will minify it down, and produce a test like:

```
from math import inf
import torch
from torch import tensor, device
import torch.fx as fxindent
import torch._dynamo
from torch._dynamo.testing import rand_strided
from torch._dynamo.debug_utils import run_fwd_maybe_bwd
from torch.nn import *
import torch._dynamo.test_case
import torch._dynamo.testing

import torch._dynamo.config
import torch._inductor.config
import torch._functorch.config
torch._dynamo.config.debug_dir_root = '.'
# torch._inductor.config.cpp.inject_relu_bug_TESTING_ONLY = 'compile_error'





class GeneratedReproTestCase(torch._dynamo.test_case.TestCase):
    def test_generated_repro_case(self):


        class Repro(torch.nn.Module):
            def __init__(self):
                super().__init__()



            def forward(self, mul):
                relu = torch.relu(mul);  mul = None
                return (relu,)


        mod = Repro()

        buf0 = torch.randn([2], dtype=torch.float32, device='cpu')  # mul 

        # TODO(voz): Add dynamic, nopython, and all sorts of other configs
        torch._dynamo.optimize('inductor')(mod)(buf0)

if __name__ == '__main__':
    from torch._dynamo.test_case import run_tests

    run_tests()
```


cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Add --produce-test flag to generate test case from minifier"


Given a file:

```
from math import inf
import torch
from torch import tensor, device
import torch.fx as fx
import torch._dynamo
from torch._dynamo.testing import rand_strided
from torch._dynamo.debug_utils import run_fwd_maybe_bwd

import torch._dynamo.config
import torch._inductor.config
import torch._functorch.config
torch._dynamo.config.debug_dir_root = '.'
torch._inductor.config.cpp.inject_relu_bug_TESTING_ONLY = 'compile_error'






from torch.nn import *
class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()



    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        sigmoid = torch.sigmoid(l_x_);  l_x_ = None
        ones = torch.ones(2)
        mul = torch.mul(sigmoid, ones);  sigmoid = ones = None
        relu = torch.relu(mul);  mul = None
        zeros = torch.zeros(2)
        add = torch.add(relu, zeros);  relu = zeros = None
        round_1 = torch.ops.aten.round(add);  add = None
        return (round_1,)


mod = Repro()

def load_args(reader):
    buf0 = reader.storage('b01abf7d0dfaf0ca12f8d2260c98f8c9b9d7bff9', 8)
    reader.tensor(buf0, (2,))  # L_x_
load_args._version = 0

if __name__ == '__main__':
    from torch._dynamo.repro.after_dynamo import run_repro
    run_repro(mod, load_args, accuracy=False, command='minify',
        save_dir='./run_2023_05_15_20_08_59_414896-pid_728954/minifier/checkpoints', autocast=False, backend='inductor')
```

If we run it with `python run_2023_05_15_20_08_59_414896-pid_728954/minifier/minifier_launcher.py minify --produce_test`

we will minify it down, and produce a test like:

```

from math import inf
import torch
from torch import tensor, device
import torch.fx as fxindent
import torch._dynamo
from torch._dynamo.testing import rand_strided
from torch._dynamo.debug_utils import run_fwd_maybe_bwd
from torch.nn import *
import torch._dynamo.test_case
import torch._dynamo.testing

import torch._dynamo.config
import torch._inductor.config
import torch._functorch.config
torch._dynamo.config.debug_dir_root = '.'
torch._inductor.config.cpp.inject_relu_bug_TESTING_ONLY = 'compile_error'





class GeneratedReproTestCase(torch._dynamo.test_case.TestCase):
    def test_generated_repro_case(self):


        class Repro(torch.nn.Module):
            def __init__(self):
                super().__init__()



            def forward(self, mul):
                relu = torch.relu(mul);  mul = None
                return (relu,)



        buf0 = torch.randn([2], dtype=torch.float32, device='cpu')  # mul 

        # TODO(voz): Add dynamic, nopython, and all sorts of other configs
        optimized_result = torch._dynamo.optimize('inductor')(Repro())(buf0)
        eager_result = Repro()(buf0)
        self.assertEqual(optimized_result, eager_result)


if __name__ == '__main__':
    from torch._dynamo.test_case import run_tests

    run_tests()

```


cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Add --produce-test flag to generate test case from minifier"


Given a file:

```
from math import inf
import torch
from torch import tensor, device
import torch.fx as fx
import torch._dynamo
from torch._dynamo.testing import rand_strided
from torch._dynamo.debug_utils import run_fwd_maybe_bwd

import torch._dynamo.config
import torch._inductor.config
import torch._functorch.config
torch._dynamo.config.debug_dir_root = '.'
torch._inductor.config.cpp.inject_relu_bug_TESTING_ONLY = 'compile_error'






from torch.nn import *
class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()



    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        sigmoid = torch.sigmoid(l_x_);  l_x_ = None
        ones = torch.ones(2)
        mul = torch.mul(sigmoid, ones);  sigmoid = ones = None
        relu = torch.relu(mul);  mul = None
        zeros = torch.zeros(2)
        add = torch.add(relu, zeros);  relu = zeros = None
        round_1 = torch.ops.aten.round(add);  add = None
        return (round_1,)


mod = Repro()

def load_args(reader):
    buf0 = reader.storage('b01abf7d0dfaf0ca12f8d2260c98f8c9b9d7bff9', 8)
    reader.tensor(buf0, (2,))  # L_x_
load_args._version = 0

if __name__ == '__main__':
    from torch._dynamo.repro.after_dynamo import run_repro
    run_repro(mod, load_args, accuracy=False, command='minify',
        save_dir='./run_2023_05_15_20_08_59_414896-pid_728954/minifier/checkpoints', autocast=False, backend='inductor')
```

If we run it with `python run_2023_05_15_20_08_59_414896-pid_728954/minifier/minifier_launcher.py minify --produce_test`

we will minify it down, and produce a test like:

```

from math import inf
import torch
from torch import tensor, device
import torch.fx as fxindent
import torch._dynamo
from torch._dynamo.testing import rand_strided
from torch._dynamo.debug_utils import run_fwd_maybe_bwd
from torch.nn import *
import torch._dynamo.test_case
import torch._dynamo.testing

import torch._dynamo.config
import torch._inductor.config
import torch._functorch.config
torch._dynamo.config.debug_dir_root = '.'
torch._inductor.config.cpp.inject_relu_bug_TESTING_ONLY = 'compile_error'





class GeneratedReproTestCase(torch._dynamo.test_case.TestCase):
    def test_generated_repro_case(self):


        class Repro(torch.nn.Module):
            def __init__(self):
                super().__init__()



            def forward(self, mul):
                relu = torch.relu(mul);  mul = None
                return (relu,)



        buf0 = torch.randn([2], dtype=torch.float32, device='cpu')  # mul 

        # TODO(voz): Add dynamic, nopython, and all sorts of other configs
        optimized_result = torch._dynamo.optimize('inductor')(Repro())(buf0)
        eager_result = Repro()(buf0)
        self.assertEqual(optimized_result, eager_result)


if __name__ == '__main__':
    from torch._dynamo.test_case import run_tests

    run_tests()

```


cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/repro/after_dynamo.py:
```diff
@@ -184,26 +185,96 @@ def generate_dynamo_fx_repro_string(
+        {textwrap.indent(model_str, ' ' * 8)}
+
+        {load_args}
+        # TODO(voz): Add dynamic, nopython, and all sorts of other configs
```
Maybe this should say flags?

ezyang commented on torch/_dynamo/debug_utils.py:
```diff
@@ -238,21 +238,34 @@ def _cuda_system_info_comment():
+    ]
+
+    formatted_code = "\n".join(lines)
+    aligned_code = textwrap.dedent(formatted_code).strip()
```
You don't need to dedent though?

voznesenskym replied:
This is a little tomfoolery to stop it from making a bunch of blank lines for when the codegen_config is empty.

I think this wanted a dedent, but perhaps that was cause of how I was using it at the callsite.

Can check...

ezyang commented:
You need to add a test for this, otherwise the code will rot faster than you can say biscuit

voznesenskym commented:
> You need to add a test for this, otherwise the code will rot faster than you can say biscuit

biscuit

Agreed tho.

ezyang commented on torch/_dynamo/repro/after_dynamo.py:
```diff
@@ -539,13 +616,23 @@ def common_flags(parser):
         args = [command, *sys.argv[1:]]
 
+    produce_test: Callable[[Any], Any] = functools.partial(
+        repro_minify, produce_test=True
```
What I was expecting to see is a `repro_produce_test`, which all it does is load up the saved FX module and prints out the test corresponding to it, and does nothing else

voznesenskym committed:
Update on "Add --produce-test flag to generate test case from minifier"


Given a file:

```
from math import inf
import torch
from torch import tensor, device
import torch.fx as fx
import torch._dynamo
from torch._dynamo.testing import rand_strided
from torch._dynamo.debug_utils import run_fwd_maybe_bwd

import torch._dynamo.config
import torch._inductor.config
import torch._functorch.config
torch._dynamo.config.debug_dir_root = '.'
torch._inductor.config.cpp.inject_relu_bug_TESTING_ONLY = 'compile_error'






from torch.nn import *
class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()



    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        sigmoid = torch.sigmoid(l_x_);  l_x_ = None
        ones = torch.ones(2)
        mul = torch.mul(sigmoid, ones);  sigmoid = ones = None
        relu = torch.relu(mul);  mul = None
        zeros = torch.zeros(2)
        add = torch.add(relu, zeros);  relu = zeros = None
        round_1 = torch.ops.aten.round(add);  add = None
        return (round_1,)


mod = Repro()

def load_args(reader):
    buf0 = reader.storage('b01abf7d0dfaf0ca12f8d2260c98f8c9b9d7bff9', 8)
    reader.tensor(buf0, (2,))  # L_x_
load_args._version = 0

if __name__ == '__main__':
    from torch._dynamo.repro.after_dynamo import run_repro
    run_repro(mod, load_args, accuracy=False, command='minify',
        save_dir='./run_2023_05_15_20_08_59_414896-pid_728954/minifier/checkpoints', autocast=False, backend='inductor')
```

If we run it with `python run_2023_05_15_20_08_59_414896-pid_728954/minifier/minifier_launcher.py minify --produce_test`

we will minify it down, and produce a test like:

```

from math import inf
import torch
from torch import tensor, device
import torch.fx as fxindent
import torch._dynamo
from torch._dynamo.testing import rand_strided
from torch._dynamo.debug_utils import run_fwd_maybe_bwd
from torch.nn import *
import torch._dynamo.test_case
import torch._dynamo.testing
import torch._dynamo.config
import torch._inductor.config
import torch._functorch.config



class GeneratedReproTestCase(torch._dynamo.test_case.TestCase):
    torch._dynamo.config.patch('debug_dir_root', '.')
    torch._inductor.config.patch('cpp.inject_relu_bug_TESTING_ONLY', 'compile_error')
    def test_generated_repro_case(self):


        class Repro(torch.nn.Module):
            def __init__(self):
                super().__init__()



            def forward(self, L_L_x_ : torch.Tensor):
                l_l_x_ = L_L_x_
                sigmoid = torch.sigmoid(l_l_x_);  l_l_x_ = None
                ones = torch.ones(2)
                mul = torch.mul(sigmoid, ones);  sigmoid = ones = None
                relu = torch.relu(mul);  mul = None
                zeros = torch.zeros(2)
                add = torch.add(relu, zeros);  relu = zeros = None
                round_1 = torch.ops.aten.round(add);  add = None
                return (round_1,)



        buf0 = torch.randn([2], dtype=torch.float32, device='cpu')  # L_L_x_ 

        # TODO(voz): Add dynamic, nopython, and all sorts of other configs
        optimized_result = torch._dynamo.optimize('inductor')(Repro())(buf0)
        eager_result = Repro()(buf0)
        self.assertEqual(optimized_result, eager_result)


if __name__ == '__main__':
    from torch._dynamo.test_case import run_tests

    run_tests()
```


cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Add --produce-test flag to generate test case from minifier"


Given a file:

```
from math import inf
import torch
from torch import tensor, device
import torch.fx as fx
import torch._dynamo
from torch._dynamo.testing import rand_strided
from torch._dynamo.debug_utils import run_fwd_maybe_bwd

import torch._dynamo.config
import torch._inductor.config
import torch._functorch.config
torch._dynamo.config.debug_dir_root = '.'
torch._inductor.config.cpp.inject_relu_bug_TESTING_ONLY = 'compile_error'






from torch.nn import *
class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()



    def forward(self, L_x_ : torch.Tensor):
        l_x_ = L_x_
        sigmoid = torch.sigmoid(l_x_);  l_x_ = None
        ones = torch.ones(2)
        mul = torch.mul(sigmoid, ones);  sigmoid = ones = None
        relu = torch.relu(mul);  mul = None
        zeros = torch.zeros(2)
        add = torch.add(relu, zeros);  relu = zeros = None
        round_1 = torch.ops.aten.round(add);  add = None
        return (round_1,)


mod = Repro()

def load_args(reader):
    buf0 = reader.storage('b01abf7d0dfaf0ca12f8d2260c98f8c9b9d7bff9', 8)
    reader.tensor(buf0, (2,))  # L_x_
load_args._version = 0

if __name__ == '__main__':
    from torch._dynamo.repro.after_dynamo import run_repro
    run_repro(mod, load_args, accuracy=False, command='minify',
        save_dir='./run_2023_05_15_20_08_59_414896-pid_728954/minifier/checkpoints', autocast=False, backend='inductor')
```

If we run it with `python run_2023_05_15_20_08_59_414896-pid_728954/minifier/minifier_launcher.py minify --produce_test`

we will minify it down, and produce a test like:

```

from math import inf
import torch
from torch import tensor, device
import torch.fx as fxindent
import torch._dynamo
from torch._dynamo.testing import rand_strided
from torch._dynamo.debug_utils import run_fwd_maybe_bwd
from torch.nn import *
import torch._dynamo.test_case
import torch._dynamo.testing
import torch._dynamo.config
import torch._inductor.config
import torch._functorch.config



class GeneratedReproTestCase(torch._dynamo.test_case.TestCase):
    torch._dynamo.config.patch('debug_dir_root', '.')
    torch._inductor.config.patch('cpp.inject_relu_bug_TESTING_ONLY', 'compile_error')
    def test_generated_repro_case(self):


        class Repro(torch.nn.Module):
            def __init__(self):
                super().__init__()



            def forward(self, L_L_x_ : torch.Tensor):
                l_l_x_ = L_L_x_
                sigmoid = torch.sigmoid(l_l_x_);  l_l_x_ = None
                ones = torch.ones(2)
                mul = torch.mul(sigmoid, ones);  sigmoid = ones = None
                relu = torch.relu(mul);  mul = None
                zeros = torch.zeros(2)
                add = torch.add(relu, zeros);  relu = zeros = None
                round_1 = torch.ops.aten.round(add);  add = None
                return (round_1,)



        buf0 = torch.randn([2], dtype=torch.float32, device='cpu')  # L_L_x_ 

        # TODO(voz): Add dynamic, nopython, and all sorts of other configs
        optimized_result = torch._dynamo.optimize('inductor')(Repro())(buf0)
        eager_result = Repro()(buf0)
        self.assertEqual(optimized_result, eager_result)


if __name__ == '__main__':
    from torch._dynamo.test_case import run_tests

    run_tests()
```


cc penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]
